src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java

```diff
@@ -267,6 +289,17 @@ public MqttClient publish(String topic, Buffer payload, MqttQoS qosLevel, boolea
 
     io.netty.handler.codec.mqtt.MqttMessage publish = MqttMessageFactory.newMessage(fixedHeader, variableHeader, buf);
 
+    switch (qosLevel) {
+      case AT_LEAST_ONCE:
+        qos1outbound.put(variableHeader.messageId(), publish);
+        countInflightQueue++;
+        break;
+      case EXACTLY_ONCE:
+        qos2outbound.put(variableHeader.messageId(), publish);
+        countInflightQueue++;
+        break;
+    }
+
     this.write(publish);
 
     if (publishSentHandler != null) {
```

Comment:
Maybe the publishSentHandler should be called here only if QoS = 0. If Qos is 1 or 2, the handler should be called when PUBACK is received (QoS 1) and PUBCOMP is received (QoS 2). It means that the publish method should not have this handler as parameter but we would need an "ad hoc" method to set it (as it happens for other handlers). It's a breaking change on the API, I know. Wdyt ?